default/python/universe_generation/monsters.py

```diff
@@ -115,62 +118,60 @@ def generate_monsters(monster_freq, systems):
         print "...", fleet_plan.name(), ": spawn rate", fleet_plan.spawn_rate(),
         print "/ spawn limit", fleet_plan.spawn_limit(),
         print "/ effective chance", basic_chance * fleet_plan.spawn_rate(),
-
-        if len(systems) < 100:
-            # Note: The WithinStarlaneJumps condition in fp.location()
-            # is the most time costly function in universe generation
-            print "/ can be spawned at", len([s for s in systems if fleet_plan.location(s)]), "systems"
-        else:
-            print  # to terminate the print line
+        fp_location_cache[fleet_plan] = set(fleet_plan.locations(systems))
+        print ("/ can be spawned at", len(fp_location_cache[fleet_plan]),
+               "of", len(systems), "systems")
         if fleet_plan.name() in nest_name_map.values():
             statistics.tracked_monsters_chance[fleet_plan.name()] = basic_chance * fleet_plan.spawn_rate()
 
     # initialize a manager for monsters that can alter the map
     # required to prevent their placement from disjoining the map
     starlane_altering_monsters = StarlaneAlteringMonsters(systems)
 
+    # collect info for tracked monster nest valid locations
+    planets = [p for s in systems for p in fo.sys_get_planets(s)]
+    tracked_nest_valid_locations = {nest: len(fo.special_locations(nest, planets))
+                                    for nest in nest_name_map}
+
     # for each system in the list that has been passed to this function, find a monster fleet that can be spawned at
     # the system and which hasn't already been added too many times, then attempt to add that monster fleet by
     # testing the spawn rate chance
+    random.shuffle(systems)
     for system in systems:
-        # collect info for tracked monster nest valid locations
-        for planet in fo.sys_get_planets(system):
-            for nest in tracked_nest_valid_locations:
-                # print "\t tracked monster check planet: %d size: %s for nest: %20s  | result: %s"
-                # % (planet, fo.planet_get_size(planet), nest, fo.special_location(nest, planet))
-                if fo.special_location(nest, planet):
-                    tracked_nest_valid_locations[nest] += 1
-
         # collect info for tracked monster valid locations
         for fp in tracked_plan_valid_locations:
-            if fp.location(system):
+            if system in fp_location_cache[fp]:
                 tracked_plan_valid_locations[fp] += 1
 
         # filter out all monster fleets whose location condition allows this system and whose counter hasn't reached 0.
-        # Note: The WithinStarlaneJumps condition in fp.location() is
-        # the most time costly function in universe generation.
         suitable_fleet_plans = [fp for fp in fleet_plans
-                                if spawn_limits[fp]
-                                and fp.location(system)
+                                if system in fp_location_cache[fp]
```

Comment:
>  I checked all of the existing monster code and monsters only care about monster in their system and distance from an empire.

Well, doesn't that mean that with each monster placed cached location conditions might become invalidated? Even is monsters care only about monsters in the same system, with that approach monsters might be placed in systems together with other monsters they are not supposed to be in the same system with.

So, either we don't pre-calculate and cache the location conditions, or we need at least inform content scripters that until we properly fix this code, they can't rely on monster fleet location conditions to work on other monsters. Existing locations conditions need to be checked and if necessary adjusted to account for that, otherwise we're going to actually introduce a bug here.

All said, I'm still extremely wary of that change. Even if it didn't cause any issues with the current monster fleet location conditions, any modification or addition that does (and it's easy enough to come up with something that does) would give unintended/buggy results.

This is a bit much for a speed gain for universe generation, even if it's a noticeable one. To cite @Morlic-fo, correctness is way more important than speed for universe generation.